src/Style.hs

```diff
@@ -16,42 +21,100 @@ data PropKey
     | BackgroundColor
   deriving (Eq, Ord, Show)
 
+data Units
+    = Px
+    | Em
+    | Percent
+    | Pt
+    deriving (Eq, Show)
+
 data PropVal
-    = Px Int
+    = NumUnit Int Units
     | Color SDL.Color
     | Auto
+    | None 
   deriving (Eq, Show)
 
--- TODO: Replace with real parser
-keyStr :: String -> PropKey
-keyStr "background-color" = BackgroundColor
-keyStr "width" = Width
+-- Parse the keys given in through a String
+parseKey :: String -> Maybe PropKey
+parseKey = either (const Nothing) Just . parseOnly (keyParser <* endOfInput) . pack 
+
+-- Parse the values given in through a String
+parseVal :: String -> Maybe PropVal
+parseVal = either (const Nothing) Just . parseOnly (valParser <* endOfInput) . pack
+
+-- Parse all possible PropKeys
+keyParser :: Parser PropKey
+keyParser = 
+     (string "background-color"     >> return BackgroundColor)
+ <|> (string "width"                >> return Width)
+ <|> (string "margin-right"         >> return MarginRight)
+ <|> (string "margin-left"          >> return MarginLeft)
+ <|> (string "padding-left"         >> return PaddingLeft)
+ <|> (string "padding-right"        >> return PaddingRight)
+ <|> (string "border-left-width"    >> return BorderLeftWidth)
+ <|> (string "border-right-width"   >> return BorderRightWidth)
+
+-- Parse the units
+unitParser :: Parser Units
+unitParser = 
+     (string "px"       >> return Px)
+ <|> (string "em"       >> return Em)
+ <|> (string "%"        >> return Percent)
+ <|> (string "pt"       >> return Pt)
+
+-- Parse the number before the units, and the units themselves
+numUnitParser :: Parser PropVal 
+numUnitParser = NumUnit <$> decimal <*> unitParser
 
-valStr :: String -> PropVal
-valStr "red" = Color (SDL.Color 255 0 0 255)
-valStr "green" = Color (SDL.Color 0 255 0 255)
-valStr "blue" = Color (SDL.Color 0 0 255 255)
-valStr s | reverse (take 2 (reverse s)) == "px" = Px (read $ reverse $ drop 2 $ reverse s)
+-- Parse a single hex character
+hexParser :: Parser Char
+hexParser = digit <|> satisfy isHex 
+                        where isHex c = inClass "abcdefABCDEF" c 
+
+-- Parser for color in #RGB format
+colorRGBParser :: Parser PropVal
+colorRGBParser = do
+    char '#'
+    r <- hexParser 
+    g <- hexParser 
+    b <- hexParser 
+    return $ Color (SDL.Color (read $ "0x" ++ [r, r]) (read $ "0x" ++ [g, g]) (read $ "0x" ++ [b, b]) 255)
+
+-- Parser for color in #RRGGBB format
+colorRRGGBBParser :: Parser PropVal
+colorRRGGBBParser = do
+    char '#'
+    r <- count 2 hexParser 
+    g <- count 2 hexParser 
+    b <- count 2 hexParser 
+    return $ Color (SDL.Color (read $ "0x" ++ r) (read $ "0x" ++ g) (read $ "0x" ++ b) 255)
```

Comment:
actually, since all of the parsers passed in will be the same, we can do something like this instead:

```language-haskell
colorHexParser :: Parser String -> Parser PropVal
colorHexParser p = Color <$> (SDL.Color <$> (char '#' *> hp) <*> hp <*> hp <*> pure 255)
  where hp = (read . ("0x" ++)) <$> p
```